lucene/core/src/test/org/apache/lucene/index/TestTermVectors.java

```diff
@@ -166,4 +118,97 @@ public void testFullMergeAddIndexesReader() throws Exception {
     verifyIndex(target);
     IOUtils.close(target, input[0], input[1]);
   }
+
+  /**
+   * Assert that a merged segment has payloads set up in fieldInfo, if at least 1 segment has
+   * payloads for this field.
+   */
+  public void testMergeWithPayloads() throws Exception {
+
+    final FieldType ft1 = new FieldType(TextField.TYPE_NOT_STORED);
+    ft1.setStoreTermVectors(true);
+    ft1.setStoreTermVectorOffsets(true);
+    ft1.setStoreTermVectorPositions(true);
+    ft1.setStoreTermVectorPayloads(true);
+    ft1.freeze();
+
+    Directory dir = newDirectory();
+    final int numDocsInSegment = 10;
+    IndexWriterConfig indexWriterConfig =
+        new IndexWriterConfig(new MockAnalyzer(random())).setMaxBufferedDocs(numDocsInSegment);
+    IndexWriter writer = new IndexWriter(dir, indexWriterConfig);
+
+    boolean hasPayloads1 = random().nextBoolean();
+    boolean hasPayloads2 = !hasPayloads1;
+    TokenStreamGenerator tkg1 = new TokenStreamGenerator(hasPayloads1);
+    TokenStreamGenerator tkg2 = new TokenStreamGenerator(hasPayloads2);
+    // create one segment with payloads, and another without payloads
+    for (int i = 0; i < numDocsInSegment; i++) {
+      Document doc = new Document();
+      doc.add(new Field("c", tkg1.newTokenStream(), ft1));
+      writer.addDocument(doc);
+    }
+    for (int i = 0; i < numDocsInSegment; i++) {
+      Document doc = new Document();
+      doc.add(new Field("c", tkg2.newTokenStream(), ft1));
+      writer.addDocument(doc);
+    }
+
+    IndexReader reader1 = writer.getReader();
+    assertEquals(
```

Comment:
Can you assert on the number of leaves?

```suggestion
    assertEquals(2, reader1.leaves().size());
    assertEquals(
```